azure/services/identities/client.go

```diff
@@ -0,0 +1,69 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package identities
+
+import (
+	"context"
+
+	"github.com/Azure/azure-sdk-for-go/services/msi/mgmt/2018-11-30/msi"
+	"github.com/Azure/go-autorest/autorest"
+	azuresdk "github.com/Azure/go-autorest/autorest/azure"
+	"sigs.k8s.io/cluster-api-provider-azure/azure"
+	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
+)
+
+// AzureClient contains the Azure go-sdk Client.
+type AzureClient struct {
+	userAssignedIdentities msi.UserAssignedIdentitiesClient
+}
+
+// NewClient creates a new MSI client from auth info.
+func NewClient(auth azure.Authorizer) *AzureClient {
+	c := newUserAssignedIdentitiesClient(auth.SubscriptionID(), auth.BaseURI(), auth.Authorizer())
+	return &AzureClient{c}
+}
+
+// newUserAssignedIdentitiesClient creates a new MSI client from subscription ID, base URI, and authorizer.
+func newUserAssignedIdentitiesClient(subscriptionID string, baseURI string, authorizer autorest.Authorizer) msi.UserAssignedIdentitiesClient {
+	userAssignedIdentitiesClient := msi.NewUserAssignedIdentitiesClientWithBaseURI(baseURI, subscriptionID)
+	azure.SetAutoRestClientDefaults(&userAssignedIdentitiesClient.Client, authorizer)
+	return userAssignedIdentitiesClient
+}
+
+// Get returns a managed service identity.
+func (ac *AzureClient) Get(ctx context.Context, resourceGroupName, name string) (msi.Identity, error) {
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "identities.AzureClient.Get")
+	defer done()
+
+	return ac.userAssignedIdentities.Get(ctx, resourceGroupName, name)
+}
+
+// GetClientID returns the client ID of a managed service identity, given its full URL identifier.
```

Comment:
how about adding an optional field for clientID in spec, and update it if it's not present? wdyt?

Reply:
That's a clever solution. I was avoiding changing the spec because of the hassle. I don't think ClientID is considered more sensitive data than the full URL (ProviderID) that we already include, so it's probably fine to have it on the spec. I'll try that out.

Reply:
@shysank wouldn't we want to add an (optional) `ClientID` field to each `UserAssignedIdentity`? Although in practice the code expects there only to be one...

Also, that would prevent the case I described above from working, where an admin could delete the MSI and recreate it but expect the ClientID to be refreshed. That may be a contrived scenario or not something we should worry about supporting. The tradeoff is making more Azure API calls.

Reply:
> where an admin could delete the MSI and recreate it but expect the ClientID to be refreshed

that definitely is a valid scenario

Another solution would be to move things around so that we get the Client ID Just-In-Time only when the cloud provider config secret doesn't exist and needs to be created. That way we don't do a GET every reconcile, only the first time when the secret is created or whenever it needs to be recreate if it was deleted (to refresh creds for example). It also optimizes things in that it doesn't require us to fetch the Client ID at all when the user is providing their own azure.json Secret and we don't need to create one.

Reply:
@mboersma I see that you added a `ClientID` field in spec, did you look into the alternative of fetching the cliend ID only if the cloud provider secret doesn't exist? I'm concerned that having the ClientID in the spec has the potential to cause redundant/confusing UX and potentially inconsistency if the user sets both but they don't point to the same identity. It could also cause issues in the case you mentioned above where the ClientID needs to be refreshed.

Reply:
> did you look into the alternative of fetching the client ID only if the cloud provider secret doesn't exist?

I did. I'm not sure it's really feasible–it boils down to the deep-equals comparison on the secret data. I moved a ton of code around trying this, but didn't see a good way to get there.

It's true that this allows the user to specify the ClientID now (and perhaps to get it wrong). But to refresh the ClientID, you can just set it to empty and we will fetch it based on the ProviderID. These are both properties on a UserAssignedIdentity, so it seemed logical to add it to that struct (even though in practice, we only ever use the first ClientID).